### ISSUE TITLE
Support for multiple bundles specified in a single config

### ISSUE THREAD

**Issue by leo** [rollup/rollup#863]
I'm currently building a pretty big application over at @zeit: While bundling the react application, I'm also bundling/transpiling the electron part.

To accomplish this, I've created a `rollup.config.electron.js` and a `rollup.config.react.js` who're both doing different things (but pretty similar). At the same time, I'm using `concurrently` to run the CLI with `-w` for both configs at the same time. In my opinion, that's all pretty hacky.

So my wish would be that I could simply combine both configs in one `rollup.config.js`:

``` js
export default [
  {
    entry: 'src/foo.js',
    dest: 'dist/foo.js'
  }, {
    entry: 'src/bar.js',
    dest: 'dist/bar.js'
  }
]
```

This has already been suggested in https://github.com/rollup/rollup/issues/703 and several approaches have been mentioned. However, not a single of them is a long-term solution, if you ask me. This could be much easier!

Because of this reason, @bahmutov created [rollem](https://github.com/bahmutov/rollem). However, the package isn't as fast as rollup itself would be because it's implementing the whole `--watch` functionality on its own. In addition, it does not support ES2015 module syntax.

IMHO, we should just build this into the core.

**Comment by TrySound** [comment]
And config will have more and more and more features and then we doing unreadable webpack clone.

**Comment by Rich-Harris** [comment]
I wasn't aware of rollem, thanks – it seems like that's a good solution for something like this. I do worry about adding too many features to the CLI, it can quickly end up looking rather hairy.

> However, the package isn't as fast as rollup itself would be because it's implementing the whole `--watch` functionality on its own. In addition, it does not support ES2015 module syntax.

Seems like both of those things could be fixed – perhaps it would involve factoring some of Rollup's internals out into packages so that things like rollem could also use them. What do you reckon @bahmutov?

**Comment by Rich-Harris** [comment]
PS @zeit is the best! Big fan here 😀  :metal:

**Comment by bahmutov** [comment]
I reckon: let's do this! 

@leo is the ES6 module syntax for the config file itself? Also, how big is the difference in speed? I am all for reusing file watch from Rollup instead of my custom solution

PS: love @zeit and just saw that you have joined them @leo - congrats!

**Comment by Rich-Harris** [comment]
> is the ES6 module syntax for the config file itself?

Rollup [rolls up the config file](https://github.com/rollup/rollup/blob/master/bin/src/runRollup.js#L61-L92) so that it can use `import` and `export`, just because it felt really weird for a Rollup config file of all things not to be a module.

The filewatching used by rollup-watch is [very basic](https://github.com/rollup/rollup-watch/blob/master/src/index.js#L10-L39) – it's just using native `fs.watch` rather than delegating to chokidar or whatever. Seems to work fine. Could be abstracted out easily enough.

**Comment by leo** [comment]
@bahmutov The speed difference is pretty big. It basically seems like it's bundling the whole thing again completely each time I make changes to a file. However, if I'm using rollup directly, it seems to take advantage of some kind of caching mechanism or it only passes the files that have changed.

If we can make rollem do this in a great way, I would definitely use it. Factoring out some of the parts related to this into packages (so that rollem can use it as well) seems like a great solution to me!

**However,** overall I'd say that implementing the array feature into the core of rollup would be much easier than all of the stuff mentioned above... Hmmm...

-

Thank you all for these kind words, btw! ❤️

**Comment by TrySound** [comment]
Some time ago we introduced `cache` option which takes previous bundle.

**Comment by snuggs** [comment]
@Rich-Harris wondering if trivially exposing an API to the watcher would solve both `rollup` / `rollem`s needs. Dependency injecting it would be a better way to go IMHO as I feel it overloads `index.js` to do too many things (besides being a module loader). This may be beyond the scope of this issue.

My $0.02

**Comment by bfred-it** [comment]
> And config will have more and more and more features and then we doing unreadable webpack clone. - @TrySound 

They need multiple bundles, which solution is clearer?

## 2 calls, 2 configs

```sh
rollup -c rollup.config.electron.js
rollup -c rollup.config.react.js
```
```js
export default {
		entry: 'src/bar.js',
		dest: 'dist/bar.js'
}
```
```js
export default {
		entry: 'src/bar.js',
		dest: 'dist/bar.js'
}
```

## 2 calls, 1 config with env

```sh
rollup -c rollup.config.js --environment BUILD:react
rollup -c rollup.config.js --environment BUILD:electron
```
```js
export default {
		entry: process.env.BUILD === 'react' ? 'src/bar.js' : 'src/foo.js',
		dest: process.env.BUILD === 'react' ? 'dist/bar.js' : 'dist/foo.js'
}
```

## 1 call, 1 config

```sh
rollup -c rollup.config.js
```
```js
export default [
	{
		entry: 'src/foo.js',
		dest: 'dist/foo.js'
	}, {
		entry: 'src/bar.js',
		dest: 'dist/bar.js'
	}
]
```

This example doesn't even show what mess it would be with 3+ bundles and extra options, while a MIMO rollup could handle any config shared across bundles as:


```js
export default [
	{
		entry: 'src/foo.js',
		dest: 'dist/foo.js'
	}, {
		entry: 'src/bar.js',
		dest: 'dist/bar.js'
	}
]
export const shared = {
	treeshaking: true
}

```

**Comment by snuggs** [comment]
Hey all. What is the status of this one? We have a pending issue dependent on the outcome of this issue. https://github.com/bahmutov/rollem/issues/7

**Comment by thisconnect** [comment]
why don't you use the rollup JS API instead of multiple configs?

**Comment by bfred-it** [comment]
@snuggs that's a good point. I think https://github.com/rollup/rollup/issues/761 would help towards that. Using the API to build multiple files would be as easy as:

```js
rollup.fromConfig({
    entry: 'src/foo.js',
    dest: 'dist/foo.js'
});
rollup.fromConfig({
    entry: 'src/bar.js',
    dest: 'dist/bar.js'
})
```

Instead of the current programmatic:

```js
rollup({
    entry: 'src/foo.js',
}).then(bundle => bundle.write({
    dest: 'dist/foo.js'
}));

rollup({
    entry: 'src/bar.js',
}).then(bundle => bundle.write({
    dest: 'dist/bar.js'
}));
```

**Comment by snuggs** [comment]
@bfred-it  have you tried rollem? Solves that bit extra I need from rollup. I'm also a contributor there as well so suggestions are welcome. Workign on a toolbelt that wraps tests around the modules and runs them in memory. A WIP but suggestions (outside of the scope of rollup/rollem) are welcome! https://github.com/snuggs/buble-tape

**Comment by TristanVALCKE** [comment]
I think @bfred-it example about 2 calls, 2 configs or 1 call, 1 config could be very interesting for multiple bundles build.

Why ? 
First that's what i'm looking for (uh uh) ! 
Second, it could be easier for new comer/new dev to write a single config file with multiples export instead of using the API.
And third, it could be use in coordination with `--environment` too, that will allow to build multiple bundles for different environments at the same time, using a single conf file.

**Comment by bfred-it** [comment]
@snuggs the problem with bubbletape is that it's opinionated. I'd rather be able to switch rollup config as needed: https://npms.io/search?q=rollup+config

Rollem doesn't support shareable configs either.

None of the three tools seem to support globbing, which starts being pretty useful once you realize that `one config file per output file` isn't manageable anymore.

---

An "aggravating" problem I think here is that the CLI has features that the API doesn't (`--watch` and `--config`) and viceversa (plugins), so whatever direction one takes, they're gonna miss some features.

Wrappers like bubbletape and Rollem are helpful, but having a flexible base tool would be a much better time investment in the long run.

How much code would be needed to replicate this line with the current rollup?

```sh
$ rollup --config node:webext --output dist --watch src/**/*.js
```

**Comment by RyanZim** [comment]
Any updates on this? I've spent a good part of my afternoon messing around, trying to get this working.

**Comment by RoboSparrow** [comment]
this seems to work for me:
```
  "devDependencies": {
     "watch": "^1.0.2"
  },
  "scripts": {
    "dev": "watch 'npm run build' ./src",
    "build": "rollup -c ./rollup.config-project.js && rollup -c ./rollup.config-demo.js",
    "test": "./node_modules/.bin/mocha --reporter spec tests"
  },

```
source: https://www.keithcirkel.co.uk/how-to-use-npm-as-a-build-tool/

caveats:
- both rollups run on any watch change
- It requires `node-watch` as devDependency, 

Unfortunately I couldn't get it to run with rollup-watch, i.e simply `"watch": "rollup -w -c ./rollup.config-project.js && rollup -w -c ./rollup.config-demo.js",` - it seems to watch only the first config

**Comment by bfred-it** [comment]
@RoboSparrow `&&` runs the command in series, not in parallel. `--watch` makes them never ending so it never gets to the second command. I think you'd need `&` or a proper task runner. The solution I suggested above works.

**Comment by djKianoosh** [comment]
fwiw, I like @bfred-it 's "2 calls, 1 config with env" approach above because in practice, rarely do I want to do both a bundle and a minified bundle at the same time. When I'm developing, I'm using the normal one, when I need to publish/deploy/etc, I will minify. So I like the flexibility of that approach. Here's how I did it with an environment config:

`rollup -c --environment minify` will use `uglify()`... 
```
import uglify from 'rollup-plugin-uglify';
let minify = process.env.minify || false
```
Then in `export default {`:
```
  dest: minify ? 'bundle.min.js' : 'bundle.js',
```
And in `plugins: [`, apply the uglify plugin or else an empty object:
```
    minify ? uglify() : {},
```

Here's the example rollup.config.js in full (dont mind the external lodash, was just playing around with externals config)
```
import nodeResolve from 'rollup-plugin-node-resolve';
import async from 'rollup-plugin-async';
import babel from 'rollup-plugin-babel';
import progress from 'rollup-plugin-progress';
import uglify from 'rollup-plugin-uglify';
import filesize from 'rollup-plugin-filesize';

let minify = process.env.minify || false

export default {
  entry: 'index.js',
  dest: minify ? 'bundle.min.js' : 'bundle.js',
  format: 'iife',
  external: [ 'lodash' ],
  globals: {
    lodash: '_'
  },
  moduleName: 'something',
  plugins: [
    nodeResolve(),
    async(),
    babel({
      exclude: 'node_modules/**'
    }),
    minify ? uglify() : {},
    progress({
      clearLine: false
    }),
    filesize()
  ]
};
```

**Comment by NeekSandhu** [comment]
For those seeking alternatives, here's my take on this problem.
Solution covers `watch`ing files and integrates seamlessly with your `gulp` setup, allowing you to have beautiful `package.json`, like 
```
'scripts': {
    'build': 'gulp build', 
    'watch': 'gulp build --watch'
} ...
```
 
```
const yargs = require('yargs').argv
const gulp = require('gulp')
const rollup = require('rollup')
const Path = require('path')

gulp.task('rollup', () => {
    const configs = [{
        entry: 'path/to/src/index.js',
        dest: 'path/to/dest/index.js',
        module: 'iife'
    }, {
        entry: 'path/another/src/index.js',
        dest: 'path/another/dest/index.js',
        module: 'cjs'
    }]

    let defaultOptsIn = {
        plugins: [],
        external: ['react-dom', 'classnames']
    }

    let defaultOptsOut = {
        sourceMap: true,
        format: 'iife',
        globals: {
            'classnames': 'classNames',
            'react-redux': 'ReactRedux',
            'redux': 'Redux'
        }
    }

    configs.forEach(config => {
        const path = `./${Path.dirname(config.entry)}/**`
        const rollItUp = () => {
            return rollup.rollup(Object.assign({},
                    defaultOptsIn, {
                        entry: entry.entry
                    }
                ))
                .then(bundle => {
                    bundle.write(Object.assign({},
                        defaultOptsOut, {
                            dest: entry.dest,
                        },
                        (entry.format) ? {
                            format: entry.format
                        } : null
                    ));
                })
        }

        gulp.src(path, rollItUp)
        if(yargs.watch)
            gulp.watch(path, rollItUp)
    })
})

gulp.task('build', ['svg', 'css', 'icons', 'rollup'])
```

**Comment by jhabdas** [comment]
> IMHO, we should just build this into the core.

Like this? https://github.com/rollup/rollup/pull/1389

/cc @bahmutov

**Comment by nathancahill** [comment]
Yes, per @leo's original issue, that's now possible.

**Comment by bfred-it** [comment]
This issue can be closed then, right @Rich-Harris?

**Comment by jhabdas** [comment]
@bfred-it perhaps @leo would like to do the honors?

**Comment by NeekSandhu** [comment]
I looked at the PR, looks beautiful, but I don't see the API documented anywhere?

**Comment by jhabdas** [comment]
@NeekSandhu please see the ChangeLog for `0.42`. If you could update the Wiki that'd be swell.

**Comment by Rich-Harris** [comment]
There's an issue to document this on rollupjs.org (https://github.com/rollup/rollupjs.org/issues/69), and if anyone has a chance to update the wiki that would be much appreciated. Basically your config file should just export an array of config objects (the config objects themselves are exactly as before).

So you can do this sort of thing, which is [how rollupjs.org is now being built](https://github.com/rollup/rollupjs.org/blob/master/rollup.config.js):

```js
import clientConfig from './client/rollup.config.js';
import serverConfig from './server/rollup.config.js';
import serviceWorkerConfig from './service-worker/rollup.config.js';

export default [
  clientConfig,
  serverConfig,
  serviceWorkerConfig
];
```

**Comment by bfred-it** [comment]
Wasn't the wiki supposed to be replaced with rollupjs? Having two separate documentation sites seems counter productive

**Comment by jhabdas** [comment]
Re: Docs. An interesting revelation. Does not two of a thing provide a means for competition and advancement? Here's an example of probably the best docs site I've seen recently. Mostly because of its deceptive simplicity, driving users deeper into Git when answers where not immediately evident.

http://patchwork2.org/

**Comment by bfred-it** [comment]
> Does not two of a thing provide a means for competition and advancement?

Not really if both are by one party; you're only forking effort, which leads to fragmented, incomplete and out of date documentation. Who'll have to update the docs on both sites? Rich, likely.

@Rich-Harris: On a related note, I think you should ask PR submitters to also update the documentation so you don't have to do all the work :)

**Comment by Grawl** [comment]
I just used @NeekSandhu's code and now I get this in my `gulp/scripts.js`:

```js
'use strict'
// global
import gulp from 'gulp'
import Path from 'path'
import {rollup} from 'rollup'
import rollupLoadPlugins from 'rollup-load-plugins'
import babelrc from 'babelrc-rollup'
import browserSync from 'browser-sync'
const $rollup = rollupLoadPlugins()
// local
import config from '../gulp-config'
let rollupCache = []
const rollupPlugins = [
	…
]
const bundleSettings = {
	sourceMap: true,
}
// public tasks
gulp.task('scripts', [], done => {
	return Promise.all(config.scriptsToCompile.map((config, index) => {
			return rollup({
				entry: config.entry,
				plugins: rollupPlugins,
				cache: rollupCache[index],
			})
				.then(bundle => {
					rollupCache[index] = bundle
					bundle.write(Object.assign(bundleSettings, {
						dest: config.dest,
						format: config.format,
						context: config.context,
						moduleName: config.moduleName,
					}))
				})
				.then(() => console.log('bundled:', config.entry))
		}))
		.then(() => console.log('all scripts bundled'))
		.then(() => browserSync.reload())
		.then(() => done)
		.catch(ex => console.error('scripts bundling error:', ex))
})
```

<details>

<summary>rollupPlugins FYI</summary>

```js
const rollupPlugins = [
	$rollup.sourcemaps({ loadMaps: true }),
	$rollup.nodeResolveExt({
		browser: true,
	}),
	$rollup.commonjs({
		include: 'node_modules/**/*',
		namedExports: { 'jquery': ['$', 'jQuery'] }
	}),
	$rollup.legacy({
		'node_modules/zoomwall.js/zoomwall.js': 'zoomwall'
	}),
	$rollup.includepaths({
		paths: [config.src],
		extensions: ['.js', '.json', '.es', '.es5', '.es6', '.es7'],
	}),
	$rollup.babel(babelrc()),
	$rollup.progress({
		clearLine: false
	}),
	$rollup.visualizer({
		filename: './rollup-visualizer.html'
	}),
]
```

</details>

**Comment by Paitum** [comment]
Why use stderr for informative messages?
https://github.com/rollup/rollup/blob/v0.47.2/bin/src/run/build.js#L42

We check for stderr output to indicate a problem in our rollup task runner.

**Comment by jhabdas** [comment]
@Paitum Have you considered asking your question on Stack Overflow or another Stack Exchange? Your question seems both off-topic for this thread.

**Comment by mnpenner** [comment]
So... which [JavaScript API](https://rollupjs.org/#javascript-api) method will accept an array of configs now? `rollup.rollup` is throwing:

> Unexpected key '0' found, expected one of: acorn, amd, banner, cache, context, entry, exports, extend, external, file, footer, format, freeze, globals, indent, input, interop, intro, legacy, moduleContext, name, noConflict, onwarn, output, outro, paths, plugins, preferConst, pureExternalModules, sourcemap, sourcemapFile, strict, targets, treeshake, watch

`rollup.watch` accepts multiple configs, but what if I want a one-off build?

**Comment by lukastaegert** [comment]
Yes, at the moment  in the case of multiple configurations, the config file parser basically calls `rollup.rollup` separately with each individual config object. If you think API consistency between `rollup.rollup` and `rollup.watch` should be improved, maybe you should open a new issue for this.

**Comment by gerardmarquinarubio** [comment]
Sorry to reopen this 6 year old issue. It's 2022, what is the best solution to have two different rollup configurations in the same file?

**Comment by Rich-Harris** [comment]
an array https://github.com/rollup/rollup/issues/863#issuecomment-306061779
